Client/Frontend/Browser/LoginsHelper.swift

```diff
@@ -35,6 +35,62 @@ class LoginsHelper: TabContentScript {
         return "loginsManagerMessageHandler"
     }
 
+    fileprivate func getPasswordOrigin(_ uriString: String, allowJS: Bool = false) -> String? {
+        var realm: String? = nil
+        if let uri = URL(string: uriString),
+            let scheme = uri.scheme, !scheme.isEmpty,
+            let host = uri.host {
+            if allowJS && scheme == "javascript" {
+                return "javascript:"
+            }
+
+            realm = "\(scheme)://\(host)"
+
+            // If the URI explicitly specified a port, only include it when
+            // it's not the default. (We never want "http://foo.com:80")
+            if let port = uri.port {
+                realm? += ":\(port)"
+            }
+        } else {
+            // bug 159484 - disallow url types that don't support a hostPort.
+            // (although we handle "javascript:..." as a special case above.)
+            log.debug("Couldn't parse origin for \(uriString)")
+            realm = nil
+        }
+        return realm
+    }
+
+    func loginRecordFromScript(_ script: [String : Any], url: URL) -> LoginRecord? {
+        guard let username = script["username"] as? String,
+            let password = script["password"] as? String else {
+                return nil
+        }
+
+        guard let origin = getPasswordOrigin(url.absoluteString) else {
+            return nil
+        }
+
+        var dict: [String : Any] = [
+            "hostname": origin,
+            "username": username,
+            "password": password
+        ]
+
+        if let formSubmit = script["formSubmitURL"] as? String {
+            dict["formSubmitURL"] = formSubmit
+        }
+
+        if let passwordField = script["passwordField"] as? String {
+            dict["passwordField"] = passwordField
+        }
+
+        if let userField = script["usernameField"] as? String {
+            dict["usernameField"] = userField
+        }
+
```

Comment:
Wouldn't an invalid dict object be handled by [LoginRecord](https://github.com/mozilla/application-services/blob/5b03ac307ffd66a1fb0ef7dd66a7afdcf1c71689/components/logins/ios/Logins/LoginRecord.swift#L123) itself? Do we need to check the fields ourself?

Reply:
yeah, we still need to currently as there's a bug in the library I'm waiting to be fixed